File: lib/utils/motion_metrics.dart
import 'dart:math' as math;

import '../models/lat_lng.dart';

class MotionMetrics {
  static double bearingDegrees(LatLng from, LatLng to) {
    final lat1 = _rad(from.latitude);
    final lat2 = _rad(to.latitude);
    final dLon = _rad(to.longitude - from.longitude);
    final y = math.sin(dLon) * math.cos(lat2);
    final x = math.cos(lat1) * math.sin(lat2) -
        math.sin(lat1) * math.cos(lat2) * math.cos(dLon);
    return (_deg(math.atan2(y, x)) + 360) % 360;
  }

  static double speedMps(LatLng from, LatLng to, double deltaSeconds) {
    if (deltaSeconds <= 0) return 0;
    final dist = _haversineM(from, to);
    return dist / deltaSeconds;
  }

  static double _haversineM(LatLng a, LatLng b) {
    const r = 6371000.0;
    final dLat = _rad(b.latitude - a.latitude);
    final dLon = _rad(b.longitude - a.longitude);
    final lat1 = _rad(a.latitude);
    final lat2 = _rad(b.latitude);
    final h = math.sin(dLat / 2) * math.sin(dLat / 2) +
        math.cos(lat1) * math.cos(lat2) * math.sin(dLon / 2) * math.sin(dLon / 2);
    return 2 * r * math.asin(math.sqrt(h));
  }

  static double _rad(double d) => d * math.pi / 180;
  static double _deg(double r) => r * 180 / math.pi;
}

File: app/src/main/java/com/cybersec/smartroute/util/MotionMetrics.java
package com.cybersec.smartroute.util;

import com.cybersec.smartroute.model.LatLng;

public final class MotionMetrics {

    private MotionMetrics() {
    }

    public static double bearingDegrees(LatLng a, LatLng b) {
        double dLon = Math.toRadians(b.longitude - a.longitude);
        double y = Math.sin(dLon) * Math.cos(Math.toRadians(b.latitude));
        double x = Math.cos(Math.toRadians(a.latitude)) * Math.sin(Math.toRadians(b.latitude))
                - Math.sin(Math.toRadians(a.latitude))
                * Math.cos(Math.toRadians(b.latitude)) * Math.cos(dLon);
        return (Math.toDegrees(Math.atan2(y, x)) + 360) % 360;
    }

    public static double speedMps(LatLng a, LatLng b, double dt) {
        if (dt <= 0) return 0;
        double km = LatLng.haversineKm(a, b);
        return (km * 1000.0) / dt;
    }
}
